### PR TITLE
ROSAENG-60614 | feat: allow partial component route updates

### PR DESCRIPTION
**Pull request by reedcort**
## PR Summary

Allow partial component route updates via `rosa edit ingress --component-routes` — users can now set a single route (e.g. only console) without providing all routes.

## Detailed Description of the Issue

The previous change (openshift/rosa#3376) followed the Classic CLI framework which required all routes to be set together (classic: oauth+console+downloads, HCP: console+downloads). The OCM API supports partial updates — setting only console leaves downloads unchanged — but the CLI rejected partial input with "the expected amount of component routes is N, but M have been supplied". After review feedback, the decision is to allow partial updates to match the API flexibility.

## Related Issues and PRs
- Jira: [ROSAENG-60614](https://redhat.atlassian.net/browse/ROSAENG-60614)
- Related PR: [openshift/rosa#3376](https://github.com/openshift/rosa/pull/3376) (merged — enabled componentRoutes for HCP)
- Related PR: [terraform-redhat/terraform-provider-rhcs#1187](https://github.com/terraform-redhat/terraform-provider-rhcs/pull/1187) (TF provider — also aligning to partial updates)

## Type of Change
- [x] feat - adds a new user-facing capability.

## Previous Behavior

`rosa edit ingress --component-routes` required all routes for the platform — classic required oauth+console+downloads (3), HCP required console+downloads (2). Passing fewer routes returned:
```
ERR: the expected amount of component routes is 2, but 1 have been supplied
```

## Behavior After This Change

Users can pass any subset of valid routes. Name validation and duplicate detection are preserved.

```bash
# Set only console (HCP)
rosa edit ingress -c mycluster --component-routes 'console: hostname=console.example.com;tlsSecretRef=console-tls' <ingress-id>

# Set only downloads (HCP)
rosa edit ingress -c mycluster --component-routes 'downloads: hostname=downloads.example.com;tlsSecretRef=downloads-tls' <ingress-id>

# Set both (still works)
rosa edit ingress -c mycluster --component-routes 'console: hostname=console.example.com;tlsSecretRef=console-tls, downloads: hostname=downloads.example.com;tlsSecretRef=downloads-tls' <ingress-id>
```

Invalid names and duplicates are still rejected:
```
ERR: 'oauth' is not a valid component name. Expected include [console, downloads]
ERR: component route "console" was supplied more than once
```

### Preconditions
- A running ROSA HCP cluster with `hypershift-component-routes` feature toggle enabled
- HyperShift operator >= v0.1.79
- TLS secret in `openshift-config` on the hosted cluster

### Test Steps
1. Set only console: `rosa edit ingress -c <cluster> --component-routes 'console: hostname=console.example.com;tlsSecretRef=console-tls' <ingress-id>`
2. Verify console route updated, downloads unchanged
3. Set only downloads: `rosa edit ingress -c <cluster> --component-routes 'downloads: hostname=downloads.example.com;tlsSecretRef=downloads-tls' <ingress-id>`
4. Verify downloads route updated, console unchanged
5. Verify invalid name still rejected: `rosa edit ingress -c <cluster> --component-routes 'oauth: hostname=x;tlsSecretRef=y' <ingress-id>`
6. Verify duplicate still rejected: `rosa edit ingress -c <cluster> --component-routes 'console: hostname=a;tlsSecretRef=b, console: hostname=c;tlsSecretRef=d' <ingress-id>`

### Expected Results
- Partial updates succeed
- Invalid names and duplicates still rejected
- Existing routes not included in the update remain unchanged on the cluster

## Proof of the Fix

```
$ go test ./cmd/edit/ingress/... -v
Ran 14 of 14 Specs in 0.002 seconds
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 0 Skipped
```

### Live cluster — partial updates and clearing

**1. Set console only:**
```
$ rosa edit ingress -c creed-hcp --component-routes 'console: hostname=console.custom.example.com;tlsSecretRef=console-tls-secret' o5n4
INFO: Updated ingress 'o5n4'
```
API state: `{"console": {"hostname": "console.custom.example.com", "tls_secret_ref": "console-tls-secret"}}`

**2. Add downloads (console unchanged):**
```
$ rosa edit ingress -c creed-hcp --component-routes 'downloads: hostname=downloads.custom.example.com;tlsSecretRef=downloads-tls-secret' o5n4
INFO: Updated ingress 'o5n4'
```
API state: `{"console": {...}, "downloads": {"hostname": "downloads.custom.example.com", "tls_secret_ref": "downloads-tls-secret"}}`

**3. Clear downloads only (console unchanged):**
```
$ rosa edit ingress -c creed-hcp --component-routes 'downloads: hostname=;tlsSecretRef=' o5n4
INFO: Updated ingress 'o5n4'
```
API state: `{"console": {"hostname": "console.custom.example.com", "tls_secret_ref": "console-tls-secret"}}`

## Breaking Changes
- [x] No breaking changes

## Developer Verification Checklist
- [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE]: <MESSAGE>`.
- [x] PR description clearly explains both **what** changed and **why**.
- [x] Relevant Jira/GitHub issues and related PRs are linked.
- [x] `make install-hooks` has been run in this clone.
- [x] Tests were added/updated where appropriate.
- [x] I manually tested the change.
- [x] `make test` passes.
- [x] `make lint` passes.
- [x] `make rosa` passes.
- [x] Documentation or repo-local agent guidance was added/updated where appropriate.
- [x] Any risk, limitation, or follow-up work is documented.

[ROSAENG-60614]: https://redhat.atlassian.net/browse/ROSAENG-60614?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **Documentation**
  * Updated the `component-routes` CLI flag help text to clarify HCP-supported components and the required `hostname` + `tlsSecretRef` pair per route.
* **Bug Fixes**
  * Improved HCP component-route validation to avoid failing solely due to a component count mismatch, while still accepting valid partial configurations.
* **Tests**
  * Expanded ingress component-route parsing tests to confirm correct output when only selected routes are provided and that empty parameter values are serialized as empty strings.

### CODE REVIEW

**Comment by amandahla** [comment]
@reedcort thanks for submitting the PR.

How is expected to delete an existing route, for example, I added console + download and now I want only console. What would be the flow?

@coderabbitai pinging you in case this already exists in the code, check. Also check if we have other commands working in a similar way (partial updates/keeps things unchanged).

**Comment by reedcort** [comment]
@amandahla Assuming both console and download routes are set to remove a single route you just need to pass it with empty values.

`rosa edit ingress -c <cluster> --component-routes 'downloads: hostname=;tlsSecretRef=' <ingress-id>`

This clears downloads while leaving console unchanged.`

File: cmd/edit/ingress/flags.go [hunk 1]
```diff
@@ -107,13 +107,6 @@ func parseComponentRoutesForAllowed(input string, allowedRoutes []string) (map[s
 	result := map[string]*cmv1.ComponentRouteBuilder{}
 	input = strings.TrimSpace(input)
 	components := strings.Split(input, ",")
-	if len(components) != len(allowedRoutes) {
```
Comment (amandahla) [inline cmd/edit/ingress/flags.go:110]: Could you update the flag message with this new behavior?

@coderabbitai suggest something.

**Comment by amandahla** [comment]
@coderabbitai do we have a e2e test for component routes? Should we update something?

**Comment by amandahla** [comment]
> @amandahla Assuming both console and download routes are set to remove a single route you just need to pass it with empty values.
> 
> `rosa edit ingress -c <cluster> --component-routes 'downloads: hostname=;tlsSecretRef=' <ingress-id>`
> 
> This clears downloads while leaving console unchanged.`

@reedcort 
Thanks for the clarification on partial updates and empty clears!

Could you paste a short before/after from a live cluster into the PR description (or a comment)? Something like `rosa describe ingress` (or the OCM/API equivalent you used) showing:

1. Both console + downloads set → update only console → downloads unchanged
2. Clear only one route with empty values → the other route unchanged

That would make the OCM merge behavior easy for reviewers to verify alongside the unit tests. Appreciate it!

**Comment by amandahla** [comment]
@coderabbitai create an issue referencing this PR
2 main points: evaluate adding a new flag  --remove-component-route and everage standard key-value map representations (--component-route key=value or JSON/YAML payloads

Details from Cursor:
UX follow-up (non-blocking for this PR): consider an explicit remove path later.
Clearing today requires knowing the internal key shape and empty-value convention:
```bash
rosa edit ingress -c <cluster> \
  --component-routes 'downloads: hostname=;tlsSecretRef=' <ingress-id>
That is hard to discover, easy to mistype, and fragile in shells because the DSL uses ; (command separator) unless the whole value is quoted.

A clearer follow-up would be something like:

rosa edit ingress <ingress-id> -c <cluster> \
  --component-route name=console,hostname=console.example.com,tlsSecretRef=console-tls
rosa edit ingress <ingress-id> -c <cluster> \
  --remove-component-route downloads

Why this is better UX than empty keys:
Intent is explicit (remove vs update), so omit ≠ delete stays unambiguous.
Users don’t need to remember hostname / tlsSecretRef just to clear a route.
Avoids shell breakage from unquoted ; / trailing = in the clear path.
Matches other ROSA edit patterns where inverse actions are explicit (e.g. boolean toggles), rather than “pass empty  values”.

**Comment by amandahla** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> set -euo pipefail
> 
> echo '=== Candidate test directories/files ==='
> fd -t f -i '(e2e|integration|acceptance|functional|test)' . \
>   | rg -i '(^|/)(e2e|integration|acceptance|functional|test)(/|_|\.|$)|ingress|route' \
>   | sort \
>   | head -300
> 
> echo '=== All component-route and ingress edit test references ==='
> rg -n -i -C 5 --glob '*_test.go' --glob '*.feature' --glob '*.yaml' --glob '*.yml' \
>   '(component-routes|component routes|ComponentRoutes|edit ingress|UpdateIngress|TlsSecretRef)' \
>   . 2>/dev/null || true
> 
> echo '=== Test framework/dependency clues ==='
> rg -n -i -C 3 --glob 'go.mod' --glob '*.go' --glob '*.feature' \
>   '(ginkgo|gomega|e2e|integration|acceptance|cucumber|godog)' \
>   . 2>/dev/null | head -400 || true
> 
> echo '=== Ingress command test layout and focused contents ==='
> ast-grep outline cmd/edit/ingress --items all 2>/dev/null || true
> sed -n '1,280p' cmd/edit/ingress/cmd_test.go 2>/dev/null || true
> 
> echo '=== Test-related CI/workflow entries ==='
> rg -n -i -C 3 --glob '.github/workflows/*' --glob 'Makefile' --glob 'makefile' \
>   '(e2e|integration|acceptance|functional)' \
>   . 2>/dev/null || true
> ```
> 
> Length of output: 57742
> 
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> set -euo pipefail
> 
> echo '=== E2E ingress test structure ==='
> ast-grep outline tests/e2e/test_rosacli_ingress.go --items all 2>/dev/null || true
> 
> echo '=== E2E ingress test: default and edit contexts ==='
> sed -n '1,540p' tests/e2e/test_rosacli_ingress.go
> 
> echo '=== CLI ingress helper definitions and use sites ==='
> rg -n -C 5 --glob '*.go' \
>   '(type .*Ingress|func \(.*\) (Edit|Describe|List|Create|Delete).*Ingress|EditIngress\(|DescribeIngress\()' \
>   tests 2>/dev/null || true
> 
> echo '=== Component-routes references specifically in executable E2E tests ==='
> rg -n -i -C 5 --glob '*.go' \
>   '(component-routes|ComponentRoutes|TlsSecretRef)' \
>   tests/e2e tests/utils/exec 2>/dev/null || true
> 
> echo '=== Ingress feature labels and E2E configuration guards ==='
> rg -n -C 4 --glob '*.go' \
>   '(Ingress|Hypershift|HCP|PrivateLink|STS)' \
>   tests/ci/labels tests/e2e/e2e_pre_check.go tests/e2e/e2e_setup_test.go tests/utils/config 2>/dev/null || true
> ```
> 
> Length of output: 51932
> 
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> set -euo pipefail
> 
> echo '=== E2E ingress reflected model and describe parsing ==='
> sed -n '52,112p' tests/utils/exec/rosacli/ingress_service.go
> 
> echo '=== Canonical component-routes describe output and assertions ==='
> ast-grep outline cmd/describe/ingress/cmd_test.go --items all 2>/dev/null || true
> sed -n '15,145p' cmd/describe/ingress/cmd_test.go
> 
> echo '=== Other E2E helpers that parse nested YAML maps ==='
> rg -n -C 6 --glob '*.go' \
>   '(YamlToMap\(\)|map\[string\].*yaml|yaml:".*Routes|yaml:".*Config)' \
>   tests/utils/exec tests/e2e 2>/dev/null | head -280 || true
> ```
> 
> Length of output: 32267
> 
> `@amandahla` Yes—there is an ingress E2E suite in `tests/e2e/test_rosacli_ingress.go`, but it currently has **no coverage for `--component-routes`**. The only coverage in this PR is parser-level in `cmd/edit/ingress/cmd_test.go`.
> 
> I recommend adding, at minimum, a command/request-level test in this PR that verifies a partial invocation—for example, only `console`—sends only the `console` entry in the ingress update payload. That directly protects the key behavior change: omitted routes must not be included or cleared by the CLI.
> 
> An E2E test would also be valuable, but needs a small amount of test infrastructure first:
> 
> * Extend `tests/utils/exec/rosacli/ingress_service.go`’s `Ingress` model to parse the nested `Component Routes` data returned by `rosa describe ingress`.
> * Add an HCP ingress scenario that:
>   
>   1. records both existing `console` and `downloads` routes,
>   2. updates only one route,
>   3. verifies that route changed while the omitted route is unchanged,
>   4. restores the original route.
> * Gate/skip it unless the E2E cluster has valid configurable component-route hostnames and TLS secret references. The current E2E profile/config does not provide those inputs.
> 
> So: **no existing E2E test; add the sparse-payload test now, and add the E2E scenario when the test environment can safely supply and restore valid route values.**

@jerichokeyne wdyt? should we add one?

File: cmd/edit/ingress/flags.go [hunk 2]
```diff
@@ -107,13 +109,6 @@ func parseComponentRoutesForAllowed(input string, allowedRoutes []string) (map[s
 	result := map[string]*cmv1.ComponentRouteBuilder{}
 	input = strings.TrimSpace(input)
 	components := strings.Split(input, ",")
```
Comment (olucasfreitas) [inline cmd/edit/ingress/flags.go:111]: Now that `componentRoutes` can be sparse, this full-map `DeepEqual` no longer matches the CLI contract. A console-only update against an ingress that already has unchanged `console` plus existing `downloads` will still PATCH and report success, because the omitted key makes the maps unequal.

File: cmd/edit/ingress/cmd_test.go [hunk 3]
```diff
@@ -117,15 +116,15 @@ var _ = Describe("Parse component routes", func() {
...
 				"oauth: hostname=oauth-host;tlsSecretRef=oauth-secret,downloads: hostname=downloads-host;tlsSecretRef=downloads-secret",
 			)
-			Expect(err).ToNot(BeNil())
-			Expect(
-				err.Error(),
-			).To(Equal("the expected amount of component routes is 3, but 2 have been supplied"))
+			Expect(err).To(BeNil())
+			Expect(result).To(HaveLen(2))
+			Expect(result).To(HaveKey("oauth"))
+			Expect(result).To(HaveKey("downloads"))
```
Comment (olucasfreitas) [inline cmd/edit/ingress/cmd_test.go:177]: These cases only show that parsing accepts fewer keys now; they don't prove the CLI sends the sparse PATCH body that this PR documents. Please add a request-level test around `run()`/`UpdateIngress` for a console-only update and an empty-value clear path.

File: cmd/edit/ingress/flags.go (hunk 1 above)
Comment (olucasfreitas) [reply cmd/edit/ingress/flags.go:110]: The latest push addresses the original ask here: the flag help now explains sparse updates, HCP key restrictions, and the empty-value clear convention. I think this specific wording thread can resolve.

**Comment by olucasfreitas** [comment]
The PR description now includes both cases this thread asked for: a console-only update that leaves downloads unchanged and a downloads clear with empty values that leaves console unchanged. That gives reviewers the backend-merge evidence needed for the behavior change.

**Comment by olucasfreitas** [comment]
I wouldn't block this PR on a new live HCP E2E, but I would block on adding one focused command/runtime PATCH assertion now. The current diff only updates parser tests, so we still need a test that proves omitted routes are not serialized and empty values do serialize as a clear.

File: cmd/edit/ingress/cmd_test.go (hunk 3 above)
Comment (reedcort) [reply cmd/edit/ingress/cmd_test.go:177]: I added two serialization-level tests that build the parsed output through `cmv1.NewIngress().ComponentRoutes().Build() ` and verify the marshaled JSON via cmv1.MarshalIngress. Is a run()-level test still needed, or are these sufficient?

**Comment by olucasfreitas** [comment]
Now that `componentRoutes` can be sparse, this full-map `DeepEqual` no longer matches the CLI contract. A console-only update against an ingress that already has unchanged `console` plus existing `downloads` will still PATCH and report success, because the omitted key makes the maps unequal.

File: cmd/edit/ingress/cmd_test.go (hunk 3 above)
Comment (olucasfreitas) [reply cmd/edit/ingress/cmd_test.go:177]: These new marshal-based cases are sufficient for the sparse/clear payload-shape concern, because `UpdateIngress` just sends the built ingress body. I wouldn't block on an extra `run()`-level test for that anymore; the remaining must-fix issue is separate in `cmd/edit/ingress/cmd.go`, where no-op detection still compares the sparse request map against the full current route map.

**Comment by olucasfreitas** [comment]
Given the new serialization tests plus the live-cluster proof in the PR description, I don't think this PR needs a new HCP E2E before merge. A follow-up E2E is still worthwhile once `tests/utils/exec/rosacli/ingress_service.go` can parse `Component Routes` from `rosa describe ingress`.

File: cmd/edit/ingress/flags.go (hunk 2 above)
Comment (reedcort) [reply cmd/edit/ingress/flags.go:111]: Addressed — added `sparseComponentRoutesEqual` to fix no-op detection for partial updates.

Verified on live cluster:

```
# Set console
$ rosa edit ingress --component-routes 'console: hostname=console.custom.example.com;tlsSecretRef=console-tls-secret' j3t8
INFO: Updated ingress 'j3t8'

# Same console again — no-op detected
$ rosa edit ingress --component-routes 'console: hostname=console.custom.example.com;tlsSecretRef=console-tls-secret' j3t8
WARN: No need to update ingress as there are no changes

# Different console value — update sent
$ rosa edit ingress --component-routes 'console: hostname=console2.custom.example.com;tlsSecretRef=console-tls-secret' j3t8
INFO: Updated ingress 'j3t8'
```
